### PR TITLE
EZP-27864: As a REST Consumer I want to remove translation from a single Version

### PR DESCRIPTION
**Pull request by alongosz**
# Implements [EZP-27864](https://jira.ez.no/browse/EZP-27864)
|  |  |
| --- | --- |
| Status:  | **Ready for a review**  |
| Target branch: | [master](https://github.com/ezsystems/ezpublish-kernel/tree/master) |
| PAPI PR:    | #2081 |
| REST endpoint: | `/content/objects/<ID>/versions/<versionNo>/translations/<languageCode>` |
| HTTP method: | `DELETE` |

This PR implements REST endpoint which allows deleting Translation, specified by language code, from a Content Object Version **Draft**.

After review and many considerations, the final version of this REST endpoint simply does the same thing as PHP API - removes Translation from a previously prepared Draft.

Constraints:
- REST API Consumer has to create Draft prior to passing it to this endpoint.
- If the removed Translation is the main one, REST API Consumer MUST update it to another one prior to using this endpoint.

It throws exceptions if:
- Version is not a draft
- User is not authorized to edit content
- The translation does not exist in the current Version
- The translation is the only one Version has or it is the main Translation.

## REST endpoint Resource Link

As the result of discussion it was brought to my attention that front-end requires REST endpoint resource link provided by `VersionInfo` View.

Endpoints that provide this information:
- `/content/objects/<contentId>/versions`
- `/content/objects/<contentId>/versions/<versionNo>`

have new property `VersionTranslationInfo` which for every language can contain `DeleteTranslation` property with `href` attribute. The property is provided only if version draft is eligible for removal. `VersionTranslationInfo` is always available on `VersionInfo` to prepare for the future feature of restoring Translation.

Sample VersionInfo REST views:
- For XML:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<VersionList media-type="application/vnd.ez.api.VersionList+xml" href="/api/ezp/v2/content/objects/61/versions">
 <VersionItem>
  <Version media-type="application/vnd.ez.api.Version+xml" href="/api/ezp/v2/content/objects/61/versions/1"/>
  <VersionInfo>
   <id>524</id>
   <versionNo>1</versionNo>
   <status>PUBLISHED</status>
   <modificationDate>2017-10-18T16:36:28+02:00</modificationDate>
   <Creator media-type="application/vnd.ez.api.User+xml" href="/api/ezp/v2/user/users/14"/>
   <creationDate>2012-09-30T12:30:00+02:00</creationDate>
   <initialLanguageCode>eng-GB</initialLanguageCode>
   <languageCodes>eng-GB</languageCodes>
   <VersionTranslationInfo media-type="application/vnd.ez.api.VersionTranslationInfo+xml">
    <Language>
     <languageCode>eng-GB</languageCode>
    </Language>
   </VersionTranslationInfo>
   <names>
    <value languageCode="eng-GB">English name</value>
   </names>
   <Content media-type="application/vnd.ez.api.ContentInfo+xml" href="/api/ezp/v2/content/objects/61"/>
  </VersionInfo>
 </VersionItem>
 <VersionItem>
  <Version media-type="application/vnd.ez.api.Version+xml" href="/api/ezp/v2/content/objects/61/versions/2"/>
  <VersionInfo>
   <id>525</id>
   <versionNo>2</versionNo>
   <status>DRAFT</status>
   <modificationDate>2017-10-18T16:36:29+02:00</modificationDate>
   <Creator media-type="application/vnd.ez.api.User+xml" href="/api/ezp/v2/user/users/14"/>
   <creationDate>2017-10-18T16:36:28+02:00</creationDate>
   <initialLanguageCode>eng-GB</initialLanguageCode>
   <languageCodes>eng-GB,pol-PL</languageCodes>
   <VersionTranslationInfo media-type="application/vnd.ez.api.VersionTranslationInfo+xml">
    <Language>
     <languageCode>eng-GB</languageCode>
    </Language>
    <Language>
     <languageCode>pol-PL</languageCode>
     <DeleteTranslation href="/api/ezp/v2/content/objects/61/versions/2/translations/pol-PL"/>
    </Language>
   </VersionTranslationInfo>
   <names>
    <value languageCode="eng-GB">English name</value>
    <value languageCode="pol-PL">Polish translated name</value>
   </names>
   <Content media-type="application/vnd.ez.api.ContentInfo+xml" href="/api/ezp/v2/content/objects/61"/>
  </VersionInfo>
 </VersionItem>
</VersionList>
``` 

- For JSON:
```json
{
    "VersionList": {
        "_media-type": "application\/vnd.ez.api.VersionList+json",
        "_href": "\/api\/ezp\/v2\/content\/objects\/60\/versions",
        "VersionItem": [
            {
                "Version": {
                    "_media-type": "application\/vnd.ez.api.Version+json",
                    "_href": "\/api\/ezp\/v2\/content\/objects\/60\/versions\/1"
                },
                "VersionInfo": {
                    "id": 522,
                    "versionNo": 1,
                    "status": "PUBLISHED",
                    "modificationDate": "2017-10-18T16:35:31+02:00",
                    "Creator": {
                        "_media-type": "application\/vnd.ez.api.User+json",
                        "_href": "\/api\/ezp\/v2\/user\/users\/14"
                    },
                    "creationDate": "2012-09-30T12:30:00+02:00",
                    "initialLanguageCode": "eng-GB",
                    "languageCodes": "eng-GB",
                    "VersionTranslationInfo": {
                        "_media-type": "application\/vnd.ez.api.VersionTranslationInfo+json",
                        "Language": [
                            {
                                "languageCode": "eng-GB"
                            }
                        ]
                    },
                    "names": {
                        "value": [
                            {
                                "_languageCode": "eng-GB",
                                "#text": "English name"
                            }
                        ]
                    },
                    "Content": {
                        "_media-type": "application\/vnd.ez.api.ContentInfo+json",
                        "_href": "\/api\/ezp\/v2\/content\/objects\/60"
                    }
                }
            },
            {
                "Version": {
                    "_media-type": "application\/vnd.ez.api.Version+json",
                    "_href": "\/api\/ezp\/v2\/content\/objects\/60\/versions\/2"
                },
                "VersionInfo": {
                    "id": 523,
                    "versionNo": 2,
                    "status": "DRAFT",
                    "modificationDate": "2017-10-18T16:35:32+02:00",
                    "Creator": {
                        "_media-type": "application\/vnd.ez.api.User+json",
                        "_href": "\/api\/ezp\/v2\/user\/users\/14"
                    },
                    "creationDate": "2017-10-18T16:35:31+02:00",
                    "initialLanguageCode": "eng-GB",
                    "languageCodes": "eng-GB,pol-PL",
                    "VersionTranslationInfo": {
                        "_media-type": "application\/vnd.ez.api.VersionTranslationInfo+json",
                        "Language": [
                            {
                                "languageCode": "eng-GB"
                            },
                            {
                                "languageCode": "pol-PL",
                                "DeleteTranslation": {
                                    "_href": "\/api\/ezp\/v2\/content\/objects\/60\/versions\/2\/translations\/pol-PL"
                                }
                            }
                        ]
                    },
                    "names": {
                        "value": [
                            {
                                "_languageCode": "eng-GB",
                                "#text": "English name"
                            },
                            {
                                "_languageCode": "pol-PL",
                                "#text": "Polish translated name"
                            }
                        ]
                    },
                    "Content": {
                        "_media-type": "application\/vnd.ez.api.ContentInfo+json",
                        "_href": "\/api\/ezp\/v2\/content\/objects\/60"
                    }
                }
            }
        ]
    }
}
```

## Action items

**TODO**:
- [x] Create REST endpoint specification
- [x] Discuss if the spec meets current capabilities of UI // as a result, changed Spec
- [x] Discuss updating main Translation / Language when needed // as the result of a discussion, REST endpoint throws exception if deleted Translation is the main one
- [x] Create functional test (probably not fully possible due to lack of Languages REST endpoints)
- [x] Finish PHP API PR #2081 and rebase
- [x] Get feedback on [Open question](#issuecomment-333491310)
- [x] Change target branch after #2081 gets merged
- [x] Implement REST endpoint
- [x] Provide resource link to this endpoint in VersionInfo REST view 
- [x] Test
- [x] Cleanup

### CODE REVIEW

File: doc/specifications/rest/REST-API-V2.rst [hunk 1]
```diff
@@ -1454,6 +1455,24 @@ Delete Content Version
     :401: If the user is not authorized to delete this version
     :403: If the version is in state published
 
+Delete Content Version Translation
+``````````````````````
+:Resource: /content/objects/<ID>/versions/<versionNo>/translations/<language_code>
+:Method: DELETE
+:Description: The translation is removed from a draft
```
Comment (dpobel) [inline doc/specifications/rest/REST-API-V2.rst]: according to https://ezsystems.app.box.com/s/d4jsidn3m0zzj39m9aei24357ajaeqfu this needs to be possible on the published version as well

File: doc/specifications/rest/REST-API-V2.rst [hunk 2]
```diff
@@ -1454,6 +1455,24 @@ Delete Content Version
...
+:Headers:
+        :X-Set-Main-Translation: [language_code] to be set as main translation (useful if deleting main translation)
```
Comment (dpobel) [inline doc/specifications/rest/REST-API-V2.rst]: and if I'm not deleting the main translation, does this still set the main translation ?
also, if that's optional, that means you need to implement a "main translation fallback" mechanism anyway.
In addition, from a PlatformUI perspective, this header will never be set unless we ask the user which translation is the main one but that's not how the feature is specified.

So I'd rather keep things "simple" by not having such header (at least in a first version).

File: doc/specifications/rest/REST-API-V2.rst (hunk 1 above)
Comment (alongosz) [reply doc/specifications/rest/REST-API-V2.rst]: > according to https://ezsystems.app.box.com/s/d4jsidn3m0zzj39m9aei24357ajaeqfu this needs to be possible on the published version as well

You're right, it's just a `remove` link in the spec, without any intermediate draft view.

I'll change it, so the REST endpoint will:
1. Check if Version is published (as box spec shows this only for published Versions).
2. Create Draft.
3. Remove language from this Draft.
4. Publish Draft w/o removed language.

@dpobel does this work for you?

File: doc/specifications/rest/REST-API-V2.rst (hunk 2 above)
Comment (alongosz) [reply doc/specifications/rest/REST-API-V2.rst]: Ok, so the only possibility here is for REST endpoint to decide on new `mainLanguageCode` if the removed language is the main one, right?

My idea here was to allow editor to make this choice if needed, but you're right - this is not within the scope of this feature.

File: doc/specifications/rest/REST-API-V2.rst [hunk 3]
```diff
@@ -1454,6 +1455,24 @@ Delete Content Version
     :401: If the user is not authorized to delete this version
     :403: If the version is in state published
 
+Delete Content Version Translation
+``````````````````````
+:Resource: /content/objects/<ID>/versions/<versionNo>/translations/<language_code>
```
Comment (dpobel) [inline doc/specifications/rest/REST-API-V2.rst]: now this is supposed to happen only on published version, do we really need to put the versionNo in the URI ? I think, we could use `currentversion` to hint that. So the resource URI would be: `/content/objects/<ID>/currentversion/translations/<language_code>`

Comment (dpobel) [reply doc/specifications/rest/REST-API-V2.rst]: Side note: if we want to follow the philosophy of our REST API, this URI should also be present somewhere in a *previous* response so that we don't have to build it manually when consuming the API. That said, I don't know where/how to do that.

Comment (alongosz) [reply doc/specifications/rest/REST-API-V2.rst]: > now this is supposed to happen only on published version, do we really need to put the versionNo in the URI ? I think, we could use currentversion to hint that. So the resource URI would be: /content/objects/<ID>/currentversion/translations/<language_code>

Ah, yes, that makes more sense. Fixed Spec in 53da041 (+ additional naming style fix)

> Side note: if we want to follow the philosophy of our REST API, this URI should also be present somewhere in a previous response so that we don't have to build it manually when consuming the API. That said, I don't know where/how to do that.

So that's why you need all these `_href`s in responses... :grinning: 
If I assume correctly that you render Version tab - that will contain this link - from the `/content/objects/<ContentId>/versions` endpoint, 
then, looking at JSON response I could add:

`VersionList.VersionItem.<PublishedVersionNo>.Version.DeleteTranslation.<languageCode>._href`

But I'm not sure yet where to add this (and if it's possible to add it where I suggested) and also if this follows proper naming conventions.
@bdunogier this seems like your area of expertise :) WDYT?

File: eZ/Publish/Core/REST/Server/Controller/Content.php [hunk 4]
```diff
@@ -736,4 +786,66 @@ protected function forward($controller)
...
+     * Get new main language code traversing through all Versions to look for the oldest one.
+     *
+     * @param \eZ\Publish\API\Repository\Values\Content\VersionInfo $versionInfo
+     * @param string $removedLanguageCode
+     *
+     * @return string
+     *
+     * @throws \eZ\Publish\Core\REST\Client\Exceptions\BadStateException
+     */
+    private function getNewMainLanguageCode(VersionInfo $versionInfo, $removedLanguageCode)
```
Comment (alongosz) [inline eZ/Publish/Core/REST/Server/Controller/Content.php]: This method implements automatic update of main Translation of a Content if it is the one we want to delete, following defined rule:

> I think we should just define a rule (and document it for the API), what makes sense to me is the chronological order. Take the oldest language introduced in the content item.

So I implemented it and this led to a very ugly example of spaghetti code and I don't think it's possible to make it better, so I prefer to get rid of it.

Simpler alternative would be to take first language from current version, however the cleanest one would be for REST endpoint not to worry about main Translation at all and throw exception in that case (should be updated via content metadata endpoint prior to deleting Translation). There's a scenario about that, but for similar feature related to Translation removal from all Versions (waiting for PM response in Jira).

// cc @andrerom @bdunogier @dpobel

**Comment by alongosz** [comment]
Note: issue with main Translation solved, it won't be handled by this REST endpoint but rather disabled for main Translation, hinting a need for update to the User.

**Comment by andrerom** [comment]
> Create functional test (probably not fully possible due to lack of Languages REST endpoints)

guess this is resolved if we more forward on the PR where .travis.yml has sql to add second language?

>  Discuss updating main Translation / Language when needed // TBD after implementation


This is done now right?

**Comment by alongosz** [comment]
> > Create functional test (probably not fully possible due to lack of Languages REST endpoints)

> guess this is resolved if we more forward on the PR where .travis.yml has sql to add second language?

Partially, for this REST endpoint I haven't finished functional test (I'm almost done). Sorry for the delay, I'm fighting flood of reviews and fixups in other issues today, so going back to working on this just now.

> >  Discuss updating main Translation / Language when needed // TBD after implementation

> This is done now right?

Yes.

**Note**: this is still work in progress as it requires #2081 to be finished and merged (I have a bad feeling that I've missed something there in the implementation, so need to go back to it as well). I've moved forward with REST endpoint to get some feedback from someone who would use this endpoint, but really should've finished PHP API PR first.

Updated list of TODOs.

**Comment by alongosz** [comment]
@bdunogier @dpobel after careful review of what REST already returns I decided that the best place to embed DeleteTranslation REST resource link is `VersionInfo`. Also preparing for the future (restoring of Versions) I've widened the scope of returned information, calling it `VersionTranslationInfo` which contains key `DeleteTranslation` only if business logic allows it (not main Translation, not single Translation, published Version only).

Every endpoint which returns `VersionInfo` should include this information. E.g.:
- `/content/objects/<contentId>/versions`
- `/content/objects/<contentId>/versions/<versionNo>`

Sample JSON output of `VersionInfo`:
```json
"VersionInfo": {
    "id": 585,
    "versionNo": 4,
    "status": "PUBLISHED",
    "modificationDate": "2017-09-21T17:27:37+02:00",
    "Creator": {
        "_media-type": "application/vnd.ez.api.User+json",
        "_href": "/api/ezp/v2/user/users/14"
    },
    "creationDate": "2017-09-21T17:27:37+02:00",
    "initialLanguageCode": "eng-GB",
    "languageCodes": "eng-GB,pol-PL",
    "VersionTranslationInfo": {
        "_media-type": "application/vnd.ez.api.VersionTranslationInfo+json",
        "Language": [
            {
                "languageCode": "eng-GB"
            },
            {
                "languageCode": "pol-PL",
                "DeleteTranslation": {
                    "_href": "/api/ezp/v2/content/objects/85/currentversion/translations/pol-PL"
                }
            }
        ]
    },
    "names": {
        "value": [
            {
                "_languageCode": "eng-GB",
                "#text": "English (UK) test"
            },
            {
                "_languageCode": "pol-PL",
                "#text": "Polish test"
            }
        ]
    },
    "Content": {
        "_media-type": "application/vnd.ez.api.ContentInfo+json",
        "_href": "/api/ezp/v2/content/objects/85"
    }
}
```

Sample XML output:
```xml
<VersionInfo>
    <id>585</id>
    <versionNo>4</versionNo>
    <status>PUBLISHED</status>
    <modificationDate>2017-09-21T17:27:37+02:00</modificationDate>
    <Creator media-type="application/vnd.ez.api.User+xml" href="/api/ezp/v2/user/users/14"/>
    <creationDate>2017-09-21T17:27:37+02:00</creationDate>
    <initialLanguageCode>eng-GB</initialLanguageCode>
    <languageCodes>eng-GB,pol-PL</languageCodes>
    <VersionTranslationInfo media-type="application/vnd.ez.api.VersionTranslationInfo+xml">
        <Language>
            <languageCode>eng-GB</languageCode>
        </Language>
        <Language>
            <languageCode>pol-PL</languageCode>
            <DeleteTranslation href="/api/ezp/v2/content/objects/85/currentversion/translations/pol-PL"/>
        </Language>
    </VersionTranslationInfo>
    <names>
        <value languageCode="eng-GB">English (UK) test</value>
        <value languageCode="pol-PL">Polish test</value>
    </names>
    <Content media-type="application/vnd.ez.api.ContentInfo+xml" href="/api/ezp/v2/content/objects/85"/>
</VersionInfo>
```
## Open question
When working with this, especially when writing functional test, it occurred to me that if REST endpoint publishes new version it might be a good idea to return its location in response header. Not sure if `Location` works for you though. Any objections? Or maybe it's not needed after all?

**Comment by alongosz** [comment]
## Open question
When working with this, especially when writing functional test, it occurred to me that since this REST endpoint publishes new version it might be a good idea to return its location in response header.

The main problem with this is that endpoint expects `HTTP DELETE` method and returns `HTTP 204 No Content` response. Looking over [RFC on HTTP `Location` header](https://tools.ietf.org/html/rfc2616#section-14.30), there's no clear statement that this header is allowed in case of `HTTP 204` (it's fine for `201` and `202`). So it's possible that for HTTP clients handling this automatically, this might lead to an unexpected behavior.

I see two possible directions to go:
1. Send `201 Created` instead of `204` and add `Location` header. However sending `Created` for `HTTP DELETE` seems to me like a strange idea.
2. Keep `204` response and use [`Content-Location` header](https://tools.ietf.org/html/rfc2616#section-14.14) for new published version.

However maybe I'm overthinking this and front-end doesn't need it?

**Comment by andrerom** [comment]
hmm, kind of gives me the feeling this should not use DELETE, however @bdunogier would know better.

File: doc/specifications/rest/REST-API-V2.rst [hunk 5]
```diff
@@ -1454,6 +1455,24 @@ Delete Content Version
     :401: If the user is not authorized to delete this version
     :403: If the version is in state published
 
+Delete Published Content Version Translation
+``````````````````````
+:Resource: /content/objects/<ID>/currentversion/translations/<languageCode>
+:Method: DELETE
+:Description: The translation is removed from a published version (creates and publishes new version)
```
Comment (bdunogier) [inline doc/specifications/rest/REST-API-V2.rst]: Nitpick: I'd rephrase that to "Removes a translation from a published version, and publishes a new version".

File: doc/specifications/rest/REST-API-V2.rst [hunk 6]
```diff
@@ -1454,6 +1455,24 @@ Delete Content Version
...
+:Response:
+
+.. code:: http
+
+    HTTP/1.1 204 No Content
+
+:Error Codes:
+        :404: if the content object or version no. or translation were not found
```
Comment (bdunogier) [inline doc/specifications/rest/REST-API-V2.rst]: Nitpick: "if the content item, version number or translation were not found".

File: eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/VersionTranslationInfo.php [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $versionInfo = $data->getVersionInfo();
+        if (empty($versionInfo->languageCodes)) {
+            return;
+        }
+
+        $generator->startObjectElement('VersionTranslationInfo');
+
+        // single language-independent conditions for deleting Translation
+        $canDelete = count($versionInfo->languageCodes) >= 2 && $versionInfo->isPublished();
```
Comment (bdunogier) [inline eZ/Publish/Core/REST/Server/Output/ValueObjectVisitor/VersionTranslationInfo.php]: I'm not a huge fan of having this calculated in this place. It would make so much more sense if the TranslationInfo object had a `canDelete` property. But I don't know where it could be done given the code flow. So I'm not asking for a change, but if you happen to have an idea... :)

**Comment by alongosz** [comment]
I also prefer more granularity and resulting flexibility. I'll change it according to your suggestions.
Note: We're deprecating `removeTranslation` and adding `deleteTranslation`, so I'll finish this one after that change.

**Comment by alongosz** [comment]
The solution got refactored according to remarks (I pushed changes yesterday).

This one is one approval away from being sent to QA (thx @bdunogier :))
Review ping @andrerom @Nattfarinn

**Comment by alongosz** [comment]
Thanks, sending to QA

**Comment by micszo** [comment]
@alongosz added comment in Jira ticket as requested (https://jira.ez.no/browse/EZP-27864?focusedCommentId=218002&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-218002).

**Comment by andrerom** [comment]
So ready for merge? :)

**Comment by alongosz** [comment]
Thanks @micszo 

> So ready for merge? :)

@andrerom yes :)
